File: concepts/07_modules_packages_and_loading/03_initialization_caching_cycles_and_dynamic_loading/cpp/test_02_failed_import_and_cache_state.cpp
// 失败初始化、缓存状态与重试。
// 共同问题：初始化失败的模块是否留在缓存；同一键再次加载是否重新执行；
// 循环依赖为何可能只看到部分初始化状态。
//
// polyglot-family: modules_packages_and_loading
// polyglot-concept: initialization_caching_cycles_and_dynamic_loading
// polyglot-related: languages/cpp/language/test_020_translation_units_odr_inline_and_modules.cpp

#include <gtest/gtest.h>

#include "support/initialization_fixture.hpp"

namespace {

TEST(TranslationUnitInitializationConcept, LinkedDefinitionInitializesOnceBeforeTestsRun) {
  const void* first = initialization_fixture::state_address();
  const void* second = initialization_fixture::state_address();

  EXPECT_EQ(initialization_fixture::construction_count(), 1);
  EXPECT_EQ(first, second);
}

TEST(TranslationUnitInitializationConcept, RuntimeModuleCacheAnalogyDoesNotApply) {
  EXPECT_EQ(initialization_fixture::construction_count(), 1);

  // translation unit 在构建期编译并由链接器组成程序；没有 Python sys.modules 或 Node
  // URL module map 可在失败后删除/重试。跨 translation unit 动态初始化依赖应改用
  // construct-on-first-use，平台 shared-library API 也不属于 C++20 import。
}

}  // namespace

File: concepts/07_modules_packages_and_loading/03_initialization_caching_cycles_and_dynamic_loading/nodejs/test_02_failed_import_and_cache_state.mjs
// 失败初始化、缓存状态与重试。
// 共同问题：初始化失败的模块是否留在缓存；同一键再次加载是否重新执行；
// 循环依赖为何可能只看到部分初始化状态。
//
// polyglot-family: modules_packages_and_loading
// polyglot-concept: initialization_caching_cycles_and_dynamic_loading
// polyglot-related: languages/nodejs/language/test_029_esm_live_bindings_namespace_dynamic_import_and_metadata.mjs

import assert from 'node:assert/strict';
import { mkdtemp, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import test from 'node:test';
import { pathToFileURL } from 'node:url';

test('失败的 ESM module record 按同一 URL 缓存 rejection', async (t) => {
  const counterKey = `__polyglot_failed_module_${process.pid}_${Date.now()}`;
  t.after(() => {
    delete globalThis[counterKey];
  });
  const source = [
    `globalThis[${JSON.stringify(counterKey)}] = (globalThis[${JSON.stringify(counterKey)}] ?? 0) + 1;`,
    "throw new Error('initialization failed');",
  ].join('\n');
  const url = `data:text/javascript,${encodeURIComponent(source)}#failed`;

  await assert.rejects(import(url), /initialization failed/);
  await assert.rejects(import(url), /initialization failed/);

  assert.equal(globalThis[counterKey], 1);

  // 与 Python 失败后删除 sys.modules 项不同，ESM module map 保留同一 URL 的失败记录；
  // 重新 import 返回 rejection 而不重新执行。改变 URL 才是不同缓存键。
});

test('循环依赖可观察尚未赋值的 var live binding', async (t) => {
  const root = await mkdtemp(join(tmpdir(), 'polyglot-cycle-'));
  t.after(() => rm(root, { recursive: true, force: true }));
  const left = join(root, 'left.mjs');
  const right = join(root, 'right.mjs');
  await writeFile(
    left,
    "import { seen } from './right.mjs';\nexport var value = 'left';\nexport const observed = seen;\n",
  );
  await writeFile(
    right,
    "import { value } from './left.mjs';\nexport const seen = value;\n",
  );

  const namespace = await import(pathToFileURL(left));

  assert.equal(namespace.value, 'left');
  assert.equal(namespace.observed, undefined);

  // var binding 在循环实例化时先初始化为 undefined；改用 const 并在初始化前读取会触发
  // TDZ ReferenceError。应通过函数延迟读取，而不是依赖部分初始化值。
});
